src/v/utils/log_hist.h

```diff
@@ -0,0 +1,184 @@
+
+public:
+    static constexpr int first_bucket_clz = std::countl_zero(
+      first_bucket_upper_bound - 1);
+    static constexpr int first_bucket_exp = 64 - first_bucket_clz;
+
+    using clock_type = std::chrono::high_resolution_clock;
+
+    /// \brief move-only type to tracking durations
+    /// if hdr_hist ptr goes out of scope, it will detach itself
+    /// and the recording will simply be ignored.
+    class measurement : public boost::intrusive::list_base_hook<> {
+    public:
+        explicit measurement(log_hist& h)
+          : _h(std::ref(h))
+          , _begin_t(log_hist::clock_type::now()) {
+            _h.get()._probes.push_back(*this);
+        }
+        measurement(const measurement&) = delete;
+        measurement& operator=(const measurement&) = delete;
+        measurement(measurement&& o) noexcept
+          : _detached(o._detached)
+          , _trace(o._trace)
+          , _h(o._h)
+          , _begin_t(o._begin_t) {
+            o.detach_hdr_hist();
+        }
+        measurement& operator=(measurement&& o) noexcept {
+            if (this != &o) {
+                this->~measurement();
+                new (this) measurement(std::move(o));
+            }
+            return *this;
+        }
+        ~measurement() noexcept {
+            if (!_detached) {
+                _h.get()._probes.erase(_h.get()._probes.iterator_to(*this));
+                // !detached && trace
+                // do not move outside of this nested if
+                if (_trace) {
+                    _h.get().record(compute_duration());
+                }
+            }
+        }
+
+        void set_trace(bool b) { _trace = b; }
+
+    private:
+        friend log_hist;
+
+        int64_t compute_duration() const {
+            return std::chrono::duration_cast<duration_t>(
+                     log_hist::clock_type::now() - _begin_t)
+              .count();
+        }
+
+        void detach_hdr_hist() { _detached = true; }
+
+        bool _detached = false;
+        bool _trace = true;
+        std::reference_wrapper<log_hist> _h;
+        log_hist::clock_type::time_point _begin_t;
+
+        friend std::ostream& operator<<(std::ostream& o, const measurement&);
+    };
+
+    std::unique_ptr<measurement> auto_measure() {
+        return std::make_unique<measurement>(*this);
+    }
+
+    log_hist()
+      : _counts(number_of_buckets) {}
+
+    /*
+     * record expects values of that are equivalent to `duration_t::count()`
+     * so make sure the input is scaled correctly.
+     */
+    void record(uint64_t val) {
+        _sample_sum += val;
+        const int i = std::min(
+          std::max(first_bucket_clz - std::countl_zero(val), 0),
+          (int)_counts.size() - 1);
+        _counts[i]++;
+    }
+
+    void record(std::unique_ptr<measurement> m) {
+        record(m->compute_duration());
+    }
+
+    seastar::metrics::histogram seastar_histogram_logform(double scale) const {
```

Comment:
`scale` is something that scales down all the samples by the given amount?

I.e., passing a scale `N` has the same effect as the alternate universe where you always did `record(sample / N)` instead of `record(sample)`?

Reply:
Yeah, that's the idea. We mainly use it in public metrics to scale from microseconds to seconds as that's the default unit Prometheus expects.